[Clang] [C++26] Expansion Statements (Part 10: Expansion Limit)

Sirraide opened the pull request:
This implements a limit on the maximum number of expansions; this isn’t in the standard (there is nothing on this in [implimits]), but my reasoning is that if someone makes a mistake and inadvertently causes e.g. 1 000 000 000 instantiations, Clang is going to hang... basically for ever (for comparison, 1 million instantiations takes about 3 seconds provided the body of the expansion statement is empty). 

For now, I’ve set the limit to `256`, but that’s probably too low. I don’t have strong opinions on the exact value; I definitely *do* think that there should be *a* limit though. We *could* in theory reuse the constexpr-steps value, but it feels a bit weird because this isn’t really doing constant evaluation so much as just template instantiation.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 10)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 10)

erichkeane reviewed (approved):
How did we derive the 'max' limit default? Experimentally, based on another limit, or in implimits?

cor3ntin reviewed (commented):
Is an actual performance concern here?

Unlike recursive things it's bound to terminate, and it's not adding recursion in itself  - the limit from the compiler perspective is the number of instantiations / objects in the declaration context.

I expect people will want to iterate over the member of `::` or `::std` and expect that to somehow work.

So i think this is fine for now but we will probably have to adjust with use cases

Sirraide commented:
> Is an actual performance concern here?
> 
> Unlike recursive things it's bound to terminate, and it's not adding recursion in itself - the limit from the compiler perspective is the number of instantiations / objects in the declaration context.
> 
> I expect people will want to iterate over the member of `::` or `::std` and expect that to somehow work.
> 
> So i think this is fine for now but we will probably have to adjust with use cases

Iirc when I was testing this it did 1 million expansions w/o hitting any existing limits; that was w/ an empty expansion statement but that already took a few seconds.

shafik reviewed (commented):
I would note, that gcc does not seem to implement a limit at least not by default. This does not mean we should not do it but wondering if they considered this at all.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 11: Final Touches and Tests) (#169690)

This makes some very minor final changes that didn’t really fit anywhere
else, and in addition it also adds the AST dump tests because I wasn’t
quite sure where else to put them (we can’t put them with the other AST
stuff since this requires Sema).
